In [1]:
%matplotlib inline

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Data Tidying and Cleaning Lab
## Reading, tidying and cleaning data. Preparing data for exploration, mining, analysis and learning

In this lab, you'll be working with the Coffee Quality Index dataset, located [here](https://www.kaggle.com/datasets/volpatto/coffee-quality-database-from-cqi). For convenience (and to save trouble in case you can't download files, or someone uploads a newer version), I've provided the dataset in the `data/` folder. The metadata (description) is at the Kaggle link. For this lab, you'll only need `merged_data_cleaned.csv`, as it is the concatenation of the other two datasets.

In this (and the following labs), you'll get several questions and problems. Do your analysis, describe it, use any tools and plots you wish, and answer. You can create any amount of cells you'd like.

Sometimes, the answers will not be unique, and they will depend on how you decide to approach and solve the problem. This is usual - we're doing science after all!

It's a good idea to save your clean dataset after all the work you've done to it.

### Problem 1. Read the dataset (1 point)
This should be self-explanatory. The first column is the index.

In [3]:
coffee_data = pd.read_csv("data/merged_data_cleaned.csv")

In [4]:
coffee_data

Unnamed: 0  Species                     Owner Country.of.Origin  \
0              0  Arabica                 metad plc          Ethiopia   
1              1  Arabica                 metad plc          Ethiopia   
2              2  Arabica  grounds for health admin         Guatemala   
3              3  Arabica       yidnekachew dabessa          Ethiopia   
4              4  Arabica                 metad plc          Ethiopia   
...          ...      ...                       ...               ...   
1334        1334  Robusta               luis robles           Ecuador   
1335        1335  Robusta               luis robles           Ecuador   
1336        1336  Robusta               james moore     United States   
1337        1337  Robusta             cafe politico             India   
1338        1338  Robusta             cafe politico           Vietnam   

                                     Farm.Name Lot.Number            Mill  \
0                                    metad plc        NaN       metad plc   
1                                    metad plc        NaN       metad plc   
2     san marcos barrancas "san cristobal cuch        NaN             NaN   
3        yidnekachew dabessa coffee plantation        NaN         wolensu   
4                                    metad plc        NaN       metad plc   
...                                        ...        ...             ...   
1334                                 robustasa   Lavado 1     our own lab   
1335                                 robustasa   Lavado 3  own laboratory   
1336                           fazenda cazengo        NaN    cafe cazengo   
1337                                       NaN        NaN             NaN   
1338                                       NaN        NaN             NaN   

             ICO.Number                                Company       Altitude  \
0             2014/2015      metad agricultural developmet plc      1950-2200   
1             2014/2015      metad agricultural developmet plc      1950-2200   
2                   NaN                                    NaN  1600 - 1800 m   
3                   NaN  yidnekachew debessa coffee plantation      1800-2200   
4             2014/2015      metad agricultural developmet plc      1950-2200   
...                 ...                                    ...            ...   
1334                NaN                              robustasa            NaN   
1335                NaN                              robustasa             40   
1336                NaN                global opportunity fund     795 meters   
1337  14-1118-2014-0087                          cafe politico            NaN   
1338                NaN                          cafe politico            NaN   

      ...       Color Category.Two.Defects           Expiration  \
0     ...       Green                    0      April 3rd, 2016   
1     ...       Green                    1      April 3rd, 2016   
2     ...         NaN                    0       May 31st, 2011   
3     ...       Green                    2     March 25th, 2016   
4     ...       Green                    2      April 3rd, 2016   
...   ...         ...                  ...                  ...   
1334  ...  Blue-Green                    1   January 18th, 2017   
1335  ...  Blue-Green                    0   January 18th, 2017   
1336  ...         NaN                    6  December 23rd, 2015   
1337  ...       Green                    1    August 25th, 2015   
1338  ...         NaN                    9    August 25th, 2015   

                      Certification.Body  \
0     METAD Agricultural Development plc   
1     METAD Agricultural Development plc   
2           Specialty Coffee Association   
3     METAD Agricultural Development plc   
4     METAD Agricultural Development plc   
...                                  ...   
1334        Specialty Coffee Association   
1335        Specialty Coffee Association   
1336        Specialty Coffee A

### Problem 2. Observations and features (1 point)
How many observations are there? How many features? Which features are numerical, and which are categorical?

**Note:** Think about the _meaning_, not the data types. The dataset hasn't been thoroughly cleaned.

In [5]:
coffee_data.shape

(1339, 44)

There is 1339 rows and 44 columns. That means 1339 observations and 44 features.

In [6]:
coffee_data.dtypes

Unnamed: 0                 int64
Species                   object
Owner                     object
Country.of.Origin         object
Farm.Name                 object
Lot.Number                object
Mill                      object
ICO.Number                object
Company                   object
Altitude                  object
Region                    object
Producer                  object
Number.of.Bags             int64
Bag.Weight                object
In.Country.Partner        object
Harvest.Year              object
Grading.Date              object
Owner.1                   object
Variety                   object
Processing.Method         object
Aroma                    float64
Flavor                   float64
Aftertaste               float64
Acidity                  float64
Body                     float64
Balance                  float64
Uniformity               float64
Clean.Cup                float64
Sweetness                float64
Cupper.Points            float64
Total.Cup.

Let's see the features that are numerical:

In [7]:
numerical_features = coffee_data.select_dtypes(include=['number']).columns.tolist()
numerical_features

['Unnamed: 0',
 'Number.of.Bags',
 'Aroma',
 'Flavor',
 'Aftertaste',
 'Acidity',
 'Body',
 'Balance',
 'Uniformity',
 'Clean.Cup',
 'Sweetness',
 'Cupper.Points',
 'Total.Cup.Points',
 'Moisture',
 'Category.One.Defects',
 'Quakers',
 'Category.Two.Defects',
 'altitude_low_meters',
 'altitude_high_meters',
 'altitude_mean_meters']

And their count:

In [8]:
len(numerical_features)

20

And this is the categorical features: 

In [9]:
categorical_features = coffee_data.select_dtypes(include=['object']).columns.tolist()
categorical_features

['Species',
 'Owner',
 'Country.of.Origin',
 'Farm.Name',
 'Lot.Number',
 'Mill',
 'ICO.Number',
 'Company',
 'Altitude',
 'Region',
 'Producer',
 'Bag.Weight',
 'In.Country.Partner',
 'Harvest.Year',
 'Grading.Date',
 'Owner.1',
 'Variety',
 'Processing.Method',
 'Color',
 'Expiration',
 'Certification.Body',
 'Certification.Address',
 'Certification.Contact',
 'unit_of_measurement']

In [10]:
coffee_data["Bag.Weight"].unique()

array(['60 kg', '1', '30 kg', '69 kg', '1 kg', '2 kg,lbs', '6', '3 lbs',
       '50 kg', '2 lbs', '100 lbs', '15 kg', '2 kg', '2', '70 kg',
       '19200 kg', '5 lbs', '1 kg,lbs', '6 kg', '0 lbs', '46 kg', '40 kg',
       '20 kg', '34 kg', '1 lbs', '660 kg', '18975 kg', '12000 kg',
       '35 kg', '66 kg', '80 kg', '132 lbs', '5 kg', '25 kg', '59 kg',
       '18000 kg', '150 lbs', '9000 kg', '18 kg', '10 kg', '29 kg',
       '1218 kg', '4 lbs', '0 kg', '13800 kg', '1500 kg', '24 kg',
       '80 lbs', '8 kg', '3 kg', '350 kg', '67 kg', '4 kg', '55 lbs',
       '100 kg', '130 lbs'], dtype=object)

We see this feature by the meaning should be numeric not categorical.

### Problem 3. Column manipulation (1 point)
Make the column names more Pythonic (which helps with the quality and... aesthetics). Convert column names to `snake_case`, i.e. `species`, `country_of_origin`, `ico_number`, etc. Try to not do it manually.

Let's see the names of the all columns:

In [11]:
coffee_data.columns

Index(['Unnamed: 0', 'Species', 'Owner', 'Country.of.Origin', 'Farm.Name',
       'Lot.Number', 'Mill', 'ICO.Number', 'Company', 'Altitude', 'Region',
       'Producer', 'Number.of.Bags', 'Bag.Weight', 'In.Country.Partner',
       'Harvest.Year', 'Grading.Date', 'Owner.1', 'Variety',
       'Processing.Method', 'Aroma', 'Flavor', 'Aftertaste', 'Acidity', 'Body',
       'Balance', 'Uniformity', 'Clean.Cup', 'Sweetness', 'Cupper.Points',
       'Total.Cup.Points', 'Moisture', 'Category.One.Defects', 'Quakers',
       'Color', 'Category.Two.Defects', 'Expiration', 'Certification.Body',
       'Certification.Address', 'Certification.Contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')

Let's implement a function that converts the names of the columns in snake case:

In [12]:
def convert_to_snake_case(column_name):
    column_name = column_name.strip().replace(" ", "_").lower()
    column_name = column_name.strip().replace(".", "_").lower()
    column_name = column_name.strip().replace(":", "").lower()
    return column_name

new_column_names = [convert_to_snake_case(col) for col in coffee_data.columns]

coffee_data.columns = new_column_names

coffee_data.columns

Index(['unnamed_0', 'species', 'owner', 'country_of_origin', 'farm_name',
       'lot_number', 'mill', 'ico_number', 'company', 'altitude', 'region',
       'producer', 'number_of_bags', 'bag_weight', 'in_country_partner',
       'harvest_year', 'grading_date', 'owner_1', 'variety',
       'processing_method', 'aroma', 'flavor', 'aftertaste', 'acidity', 'body',
       'balance', 'uniformity', 'clean_cup', 'sweetness', 'cupper_points',
       'total_cup_points', 'moisture', 'category_one_defects', 'quakers',
       'color', 'category_two_defects', 'expiration', 'certification_body',
       'certification_address', 'certification_contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')

### Problem 4. Bag weight (1 point)
What's up with the bag weights? Make all necessary changes to the column values. Don't forget to document your methods and assumptions.

Now let's see the unique values of the bag_weight feature:

In [41]:
coffee_data["bag_weight"].unique()

array(['60 kg', '1', '30 kg', '69 kg', '1 kg', '2 kg,lbs', '6', '3 lbs',
       '50 kg', '2 lbs', '100 lbs', '15 kg', '2 kg', '2', '70 kg',
       '19200 kg', '5 lbs', '1 kg,lbs', '6 kg', '0 lbs', '46 kg', '40 kg',
       '20 kg', '34 kg', '1 lbs', '660 kg', '18975 kg', '12000 kg',
       '35 kg', '66 kg', '80 kg', '132 lbs', '5 kg', '25 kg', '59 kg',
       '18000 kg', '150 lbs', '9000 kg', '18 kg', '10 kg', '29 kg',
       '1218 kg', '4 lbs', '0 kg', '13800 kg', '1500 kg', '24 kg',
       '80 lbs', '8 kg', '3 kg', '350 kg', '67 kg', '4 kg', '55 lbs',
       '100 kg', '130 lbs'], dtype=object)

Now let's convert pounds to kilograms. Iterating through each row in the DataFrame handle cases where there's only one element (like '1' or '6') or cases where there are two elements ( 'kg' and 'lbs') for one value:

In [47]:
for idx, col in enumerate(coffee_data["bag_weight"]):
    elements = col.split()
    num = elements[0]
    
    if len(elements) < 2:
        elements.append('kg')  
    
    if len(elements) >= 2:
        kg = elements[1]
        if kg == 'lbs':
            converted_weight = lbs_to_kg(num)
            elements[0] = f"{converted_weight:.2f}"  
            elements[1] = 'kg'

        coffee_data.at[idx, 'bag_weight'] = " ".join(elements)

Below we see the applied changes:

In [48]:
coffee_data.bag_weight.unique()

array(['60 kg', '1 kg', '30 kg', '69 kg', '2 kg,lbs', '6 kg', '1.36 kg',
       '50 kg', '0.91 kg', '45.36 kg', '15 kg', '2 kg', '70 kg',
       '19200 kg', '2.27 kg', '1 kg,lbs', '0.00 kg', '46 kg', '40 kg',
       '20 kg', '34 kg', '0.45 kg', '660 kg', '18975 kg', '12000 kg',
       '35 kg', '66 kg', '80 kg', '59.87 kg', '5 kg', '25 kg', '59 kg',
       '18000 kg', '68.04 kg', '9000 kg', '18 kg', '10 kg', '29 kg',
       '1218 kg', '1.81 kg', '0 kg', '13800 kg', '1500 kg', '24 kg',
       '36.29 kg', '8 kg', '3 kg', '350 kg', '67 kg', '4 kg', '24.95 kg',
       '100 kg', '58.97 kg'], dtype=object)

### Problem 5. Dates (1 point)
This should remind you of problem 4 but it's slightly nastier. Fix the harvest years, document the process.

While you're here, fix the expiration dates, and grading dates. Unlike the other column, these should be dates (`pd.to_datetime()` is your friend).

### Problem 6. Countries (1 point)
How many coffees are there with unknown countries of origin? What can you do about them?

### Problem 7. Owners (1 point)
There are two suspicious columns, named `Owner`, and `Owner.1` (they're likely called something different after you solved problem 3). Do something about them. Is there any link to `Producer`?

### Problem 8. Coffee color by country and continent (1 point)
Create a table which shows how many coffees of each color are there in every country. Leave the missing values as they are.

**Note:** If you ask me, countries should be in rows, I prefer long tables much better than wide ones.

Now do the same for continents. You know what continent each country is located in.

### Problem 9. Ratings (1 point)
The columns `Aroma`, `Flavor`, etc., up to `Moisture` represent subjective ratings. Explore them. Show the means and range; draw histograms and / or boxplots as needed. You can even try correlations if you want. What's up with all those ratings?

### Problem 10. High-level errors (1 point)
Check the countries against region names, altitudes, and companies. Are there any discrepancies (e.g. human errors, like a region not matching the country)? Take a look at the (cleaned) altitudes; there has been a lot of preprocessing done to them. Was it done correctly?

### * Problem 11. Clean and explore at will
The dataset claimed to be clean, but we were able to discover a lot of things to fix and do better.

Play around with the data as much as you wish, and if you find variables to tidy up and clean - by all means, do that!